crypto/merklekeystore/keystore_test.go

```diff
@@ -0,0 +1,538 @@
+	hashable, sig := makeSig(signer, start, a)
+
+	a.Error(signer.GetVerifier().Verify(0, start, 1, hashable, sig))
+	a.Error(signer.GetVerifier().Verify(start, start+1, 1, hashable, sig))
+
+	hashable, sig = makeSig(signer, start+1, a)
+	a.Error(signer.GetVerifier().Verify(start, start, 1, hashable, sig))
+	a.Error(signer.GetVerifier().Verify(start, start+2, 1, hashable, sig))
+}
+
+func TestBadMerkleProofInSignature(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	a := require.New(t)
+
+	start, _, signer := generateTestSignerAux(a)
+	defer signer.keyStore.store.Close()
+	hashable, sig := makeSig(signer, start, a)
+
+	sig2 := copySig(sig)
+	sig2.Proof.Path = sig2.Proof.Path[:len(sig2.Proof.Path)-1]
+	a.Error(signer.GetVerifier().Verify(start, start, 1, hashable, sig2))
+
+	sig3 := copySig(sig)
+	someDigest := crypto.Digest{}
+	rand.Read(someDigest[:])
+	sig3.Proof.Path[0] = someDigest[:]
+	a.Error(signer.GetVerifier().Verify(start, start, 1, hashable, sig3))
+}
+
+func copySig(sig Signature) Signature {
+	bsig := make([]byte, len(sig.ByteSignature))
+	copy(bsig, sig.ByteSignature)
+
+	return Signature{
+		ByteSignature: bsig,
+		Proof:         copyProof(sig.Proof),
+		VerifyingKey:  sig.VerifyingKey,
+	}
+}
+
+func copyProof(proof Proof) Proof {
+	path := make([]crypto.GenericDigest, len(proof.Path))
+	for i, digest := range proof.Path {
+		path[i] = make([]byte, len(digest))
+		copy(path[i], digest)
+	}
+	return Proof{
+		Path:        path,
+		HashFactory: proof.HashFactory,
+	}
+}
+
+func TestIncorrectByteSignature(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	a := require.New(t)
+
+	start, _, signer := generateTestSignerAux(a)
+	defer signer.keyStore.store.Close()
+	hashable, sig := makeSig(signer, start, a)
+
+	sig2 := sig
+	bs := make([]byte, len(sig.ByteSignature))
+	copy(bs, sig2.ByteSignature)
+	bs[0]++
+	sig2.ByteSignature = bs
+	a.Error(signer.GetVerifier().Verify(start, start, 1, hashable, sig2))
+}
+
+func TestAttemptToUseDifferentKey(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	a := require.New(t)
+	var err error
+
+	start, _, signer := generateTestSignerAux(a)
+	defer signer.keyStore.store.Close()
+	hashable, sig := makeSig(signer, start+1, a)
+	// taking signature for specific round and changing the round
+
+	// taking signature and changing the key to match different round
+	sig2 := sig
+	key, err := signer.keyStore.GetKey(start)
+	a.NoError(err)
+
+	sig2.VerifyingKey = *(key.GetSigner().GetVerifyingKey())
+	a.Error(signer.GetVerifier().Verify(start, start+1, 1, hashable, sig2))
+}
+
+func TestMarshal(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	a := require.New(t)
+
+	signer := generateTestSigner(crypto.FalconType, 0, 10, 1, a)
+	store := signer.keyStore.store
+	defer store.Close()
+
+	out := protocol.Encode(signer)
+	decodeInto := &Signer{}
+	a.NoError(protocol.Decode(out, decodeInto))
+	decodeInto.keyStore.store = store // restore PersistentKeystore
+	a.Equal(signer, decodeInto)
```

Comment:
`Equal` first argument is the expected value, second is the checked value. The parameter order must be reversed.